src/doorman/Channel.cpp

```diff
@@ -0,0 +1,393 @@
+                    log_out("[%d] NAWS = %dx%d", id_, width_, height_);
+                    break;
+                }
+
+                /* Remember eob is 1 byte behind the end of buffer */
+                memmove(ptr, p + 2, nLeft - ((p + 2) - ptr));
+                nLeft -= ((p + 2) - ptr);
+
+                break;
+            }
+
+            default:
+                memmove(ptr, ptr + 2, nLeft - 2);
+                nLeft -= 2;
+                break;
+            }
+
+        } else {
+            ptr++;
+            nRealBytes++;
+            nLeft--;
+        }
+    }
+    /* Now to update the buffer stats */
+    buffer_.resize(nRealBytes);
+
+    /* Second pass - look for whole lines of text */
+    nLeft = buffer_.size();
+    byte *ptr;
+    for (ptr = buffer_.data(); nLeft; ++ptr, --nLeft) {
+        char c;
+        switch (*ptr) {
+        case '\n':
+        case '\r':
+            // TODO this is terribly ghettish. don't need to memmove, and all that. ugh. But write tests first...
+            // then change...
+            c = *ptr; // legacy<-
+            /* Send it to the MUD */
+            mud_->on_client_message(
+                *this, std::string_view(reinterpret_cast<const char *>(buffer_.data()), ptr - buffer_.data()));
+            /* Check for \n\r or \r\n */
+            if (nLeft > 1 && (*(ptr + 1) == '\r' || *(ptr + 1) == '\n') && *(ptr + 1) != c) {
+                memmove(buffer_.data(), ptr + 2, nLeft - 2);
+                nLeft--;
+            } else if (nLeft) {
+                memmove(buffer_.data(), ptr + 1, nLeft - 1);
+            } // else nLeft is zero, so we might as well avoid calling memmove(x,y,
+            // 0)
+
+            /* The +1 at the top of the loop resets ptr to buffer */
+            ptr = buffer_.data() - 1;
+            break;
+        }
+    }
+    buffer_.resize(ptr - buffer_.data());
+    return true;
+}
+
+void Channel::incomingHostnameInfo() {
+    log_out("[%d on %d] Incoming IdentD info", id_, host_lookup_fd_.number());
+    // Move out the fd, so we know however we exit from this point it will be closed.
+    auto response_fd = std::move(host_lookup_fd_);
+    try {
+        auto numBytes = response_fd.read_all<size_t>();
+        char host_buffer[1024];
+        if (numBytes > sizeof(host_buffer)) {
+            log_out("Ident responded with > %lu bytes - possible spam attack", sizeof(host_buffer));
+            numBytes = sizeof(host_buffer);
+        }
+        response_fd.read_all(host_buffer, numBytes);
+        hostname_ = std::string_view(host_buffer, numBytes);
+
+    } catch (const std::runtime_error &re) {
+        log_out("[%d] IdentD pipe died on read: %s", id_, re.what());
+        host_lookup_pid_ = 0;
+        return;
+    }
+    // Log the source IP but mask it for privacy.
+    log_out("[%d] %s", id_, get_masked_hostname(hostname_).c_str());
+    sendInfoPacket();
+}
+
+void Channel::sendInfoPacket() const {
+    char buffer[4096];
```

Comment:
inconsistent buffer size here compared to MaxIncoming (related to size comment I made earlier today too but we can tackle those separately)

Reply:
Right. I think a `Packet::MaxPayloadSize` on both sides is probably needed. Will add a TODO in the code and in the PR notes above to cover this.